// @zilliz/milvus2-sdk-node/dist/milvus/http/Collection.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Collection = void 0;
function Collection(Base) {
    return class extends Base {
        // POST create collection
        createCollection(data) {
            return __awaiter(this, void 0, void 0, function* () {
                const url = `/vector/collections/create`;
                return yield this.POST(url, data);
            });
        }
        // GET describe collection
        describeCollection(params) {
            return __awaiter(this, void 0, void 0, function* () {
                const url = `/vector/collections/describe`;
                return yield this.GET(url, { params });
            });
        }
        // POST drop collection
        dropCollection(data) {
            return __awaiter(this, void 0, void 0, function* () {
                const url = `/vector/collections/drop`;
                return yield this.POST(url, data);
            });
        }
        // GET list collections
        listCollections(params = {}) {
            return __awaiter(this, void 0, void 0, function* () {
                const url = `/vector/collections`;
                return yield this.GET(url, { params });
            });
        }
    };
}
exports.Collection = Collection;
