// @zilliz/milvus2-sdk-node/dist/milvus/http/Vector.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Vector = void 0;
function Vector(Base) {
    return class extends Base {
        // GET get data
        get(params) {
            return __awaiter(this, void 0, void 0, function* () {
                const url = `/vector/get`;
                return yield this.GET(url, { params });
            });
        }
        // POST insert data
        insert(data) {
            return __awaiter(this, void 0, void 0, function* () {
                const url = `/vector/insert`;
                return yield this.POST(url, data);
            });
        }
        // POST query data
        query(data) {
            return __awaiter(this, void 0, void 0, function* () {
                const url = `/vector/query`;
                return yield this.client.post(url, data);
            });
        }
        // POST search data
        search(data) {
            return __awaiter(this, void 0, void 0, function* () {
                const url = `/vector/search`;
                return yield this.POST(url, data);
            });
        }
        // POST delete collection
        delete(data) {
            return __awaiter(this, void 0, void 0, function* () {
                const url = `/vector/delete`;
                return yield this.POST(url, data);
            });
        }
    };
}
exports.Vector = Vector;
