// @ironclad/rivet-core/dist/esm/model/nodes/IfElseNode.js
import { NodeImpl } from '../NodeImpl.js';
import { nodeDefinition } from '../NodeDefinition.js';
import {} from '../NodeBase.js';
import {} from '../DataValue.js';
import { nanoid } from 'nanoid/non-secure';
import { ControlFlowExcludedPort } from '../../utils/symbols.js';
import { dedent } from 'ts-dedent';
export class IfElseNodeImpl extends NodeImpl {
    static create = () => {
        const chartNode = {
            type: 'ifElse',
            title: 'If/Else',
            id: nanoid(),
            data: {},
            visualData: {
                x: 0,
                y: 0,
                width: 175,
            },
        };
        return chartNode;
    };
    getInputDefinitions() {
        return [
            {
                id: 'if',
                title: 'If',
                dataType: 'any',
            },
            {
                id: 'true',
                title: 'True',
                dataType: 'any',
            },
            {
                id: 'false',
                title: 'False',
                dataType: 'any',
            },
        ];
    }
    getOutputDefinitions() {
        return [
            {
                id: 'output',
                title: 'Output',
                dataType: 'any',
            },
        ];
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Takes in three inputs: a condition, a true value, and a false value. If the condition is truthy, the true value is passed through the output port. If the condition is not truthy, the false value is passed through the output port.

        This node can "consume" a \`Not Ran\` to continue a graph from that point.
      `,
            infoBoxTitle: 'If/Else Node',
            contextMenuTitle: 'If/Else',
            group: ['Logic'],
        };
    }
    async process(inputData) {
        const ifValue = inputData['if'];
        const trueValue = inputData['true'] ?? { type: 'any', value: undefined };
        const falseValue = inputData['false'] ?? { type: 'any', value: undefined };
        if (!(trueValue || falseValue)) {
            return {
                ['output']: {
                    type: 'control-flow-excluded',
                    value: undefined,
                },
            };
        }
        if (ifValue?.type === 'control-flow-excluded') {
            return {
                ['output']: falseValue,
            };
        }
        if (inputData[ControlFlowExcludedPort]) {
            return {
                ['output']: falseValue,
            };
        }
        if (ifValue?.value == null) {
            return {
                ['output']: falseValue,
            };
        }
        if (ifValue?.type && ifValue.type === 'boolean') {
            return {
                ['output']: ifValue.value ? trueValue : falseValue,
            };
        }
        if (ifValue?.type === 'string') {
            return {
                ['output']: ifValue.value.length > 0 ? trueValue : falseValue,
            };
        }
        if (ifValue?.type === 'chat-message') {
            return {
                ['output']: ifValue.value.message.length > 0 ? trueValue : falseValue,
            };
        }
        if (ifValue?.type.endsWith('[]')) {
            return {
                ['output']: ifValue.value.length > 0 ? trueValue : falseValue,
            };
        }
        if (ifValue?.type === 'any' || ifValue?.type === 'object') {
            return {
                ['output']: !!ifValue.value ? trueValue : falseValue,
            };
        }
        return {
            ['output']: falseValue,
        };
    }
}
export const ifElseNode = nodeDefinition(IfElseNodeImpl, 'If/Else');
