// @ironclad/rivet-core/dist/esm/model/nodes/IfNode.js
import { NodeImpl } from '../NodeImpl.js';
import { nodeDefinition } from '../NodeDefinition.js';
import {} from '../NodeBase.js';
import {} from '../DataValue.js';
import { nanoid } from 'nanoid/non-secure';
import { dedent } from 'ts-dedent';
export class IfNodeImpl extends NodeImpl {
    static create = () => {
        const chartNode = {
            type: 'if',
            title: 'If',
            id: nanoid(),
            data: {},
            visualData: {
                x: 0,
                y: 0,
                width: 125,
            },
        };
        return chartNode;
    };
    getInputDefinitions() {
        return [
            {
                id: 'if',
                title: 'If',
                dataType: 'string',
            },
            {
                id: 'value',
                title: 'Value',
                dataType: 'string',
            },
        ];
    }
    getOutputDefinitions() {
        return [
            {
                id: 'output',
                title: 'True',
                dataType: 'string',
            },
            {
                id: 'falseOutput',
                title: 'False',
                dataType: 'string',
            },
        ];
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Takes in a condition and a value. If the condition is truthy, the value is passed through the True port, and the False port is not ran.
        If the condition is falsy, the value is passed through the False port, and the True port is not ran.
      `,
            infoBoxTitle: 'If Node',
            contextMenuTitle: 'If',
            group: ['Logic'],
        };
    }
    async process(inputData) {
        const ifValue = inputData['if'];
        const value = inputData['value'] ?? { type: 'any', value: undefined };
        const isFalse = {
            output: {
                type: 'control-flow-excluded',
                value: undefined,
            },
            falseOutput: value,
        };
        if (!ifValue) {
            return isFalse;
        }
        if (ifValue.type === 'control-flow-excluded') {
            return isFalse;
        }
        if (ifValue.type === 'string' && !ifValue.value) {
            return isFalse;
        }
        if (ifValue.type === 'boolean' && !ifValue.value) {
            return isFalse;
        }
        if (ifValue.type.endsWith('[]')) {
            const value = ifValue;
            if (!value.value || value.value.length === 0) {
                return isFalse;
            }
        }
        return {
            ['output']: value,
            ['falseOutput']: {
                type: 'control-flow-excluded',
                value: undefined,
            },
        };
    }
}
export const ifNode = nodeDefinition(IfNodeImpl, 'If');
