// @ironclad/rivet-core/dist/esm/model/nodes/ReadFileNode.js
import {} from '../NodeBase.js';
import {} from '../DataValue.js';
import { NodeImpl } from '../NodeImpl.js';
import { nodeDefinition } from '../NodeDefinition.js';
import { nanoid } from 'nanoid/non-secure';
import { expectType } from '../../utils/index.js';
import {} from '../ProcessContext.js';
import { dedent } from 'ts-dedent';
export class ReadFileNodeImpl extends NodeImpl {
    static create() {
        return {
            id: nanoid(),
            type: 'readFile',
            title: 'Read File',
            visualData: { x: 0, y: 0, width: 250 },
            data: {
                path: '',
                usePathInput: true,
                errorOnMissingFile: false,
            },
        };
    }
    getInputDefinitions() {
        const inputDefinitions = [];
        if (this.chartNode.data.usePathInput) {
            inputDefinitions.push({
                id: 'path',
                title: 'Path',
                dataType: 'string',
                coerced: false,
            });
        }
        return inputDefinitions;
    }
    getOutputDefinitions() {
        return [
            {
                id: 'content',
                title: 'Content',
                dataType: 'string',
            },
        ];
    }
    static getUIData() {
        return {
            infoBoxBody: dedent `
        Reads the contents of the specified file and outputs it as a string.
      `,
            infoBoxTitle: 'Read File Node',
            contextMenuTitle: 'Read File',
            group: ['Input/Output'],
        };
    }
    async process(inputData, context) {
        const { nativeApi } = context;
        if (nativeApi == null) {
            throw new Error('This node requires a native API to run.');
        }
        const path = this.chartNode.data.usePathInput
            ? expectType(inputData['path'], 'string')
            : this.chartNode.data.path;
        try {
            const content = await nativeApi.readTextFile(path, undefined);
            return {
                ['content']: { type: 'string', value: content },
            };
        }
        catch (err) {
            if (this.chartNode.data.errorOnMissingFile) {
                throw err;
            }
            else {
                return {
                    ['content']: { type: 'string', value: '(file does not exist)' },
                };
            }
        }
    }
}
export const readFileNode = nodeDefinition(ReadFileNodeImpl, 'Read File');
